backend/hct_mis_api/one_time_scripts/program_cycle_data_migration.py

```diff
@@ -0,0 +1,173 @@
+        if new_hh_ids.intersection(hh_ids_in_cycles):
+            cycle = create_new_program_cycle(
+                str(payment_plan.program_id), ProgramCycle.ACTIVE, payment_plan_start_date, payment_plan_end_date
+            )
+            break
+    # update TP
+    TargetPopulation.objects.filter(id=payment_plan.target_population_id).update(program_cycle=cycle)
+    # update Payment Plan
+    PaymentPlan.objects.filter(id=payment_plan.id).update(program_cycle=cycle)
+
+    # adjust cycle start and end dates based on PaymentPlan
+    # if cycle start date if after than payment plan
+
+    cycle_start_date = (
+        datetime.strptime(cycle.start_date, "%Y-%m-%d").date()
+        if isinstance(cycle.start_date, str)
+        else cycle.start_date
+    )
+    cycle_end_date = (
+        datetime.strptime(cycle.end_date, "%Y-%m-%d").date() if isinstance(cycle.end_date, str) else cycle.end_date
+    )
+
+    if cycle_start_date and cycle_start_date > payment_plan.start_date.date():
+        cycle.start_date = payment_plan.start_date.date()
+    # if cycle end date is before payment plan end date
+    if (cycle_end_date and cycle_end_date < payment_plan.end_date.date()) or not cycle_end_date:
+        cycle.end_date = payment_plan.end_date.date()
+    if cycle.status != ProgramCycle.ACTIVE:
+        cycle.status = ProgramCycle.ACTIVE
+    cycle.save(update_fields=["start_date", "end_date", "status"])
+
+
+def program_cycle_data_migration(batch_size: int = 1000) -> None:
+    start_time = timezone.now()
+    logger.info("Hi There! Started Program Cycle Data Migration.")
+    logger.info(f"Cycles before running creation: {ProgramCycle.objects.all().count()}")
+    for ba in BusinessArea.objects.all().only("id", "name"):
+        logger.info(f"Started processing {ba.name}...")
+
+        for program in Program.objects.filter(business_area_id=ba.id).only(
+            "id", "name", "start_date", "end_date", "status"
+        ):
+            logger.info(f"** Creating Program Cycles for program {program.name}")
+
+            payment_plan_qs = PaymentPlan.objects.filter(program_id=program.id).only("id")
+            pp_count = payment_plan_qs.count()
+            pp_ids = list(payment_plan_qs.values_list("id", flat=True))
+
+            for batch_start in range(0, pp_count, batch_size):
+                batch_end = batch_start + batch_size
+                logger.info(f"Handling {batch_start} - {batch_end}/{pp_count} PaymentPlans")
+
+                processing_ids = pp_ids[batch_start:batch_end]
+                for pp_id in processing_ids:
+                    payment_plan = PaymentPlan.objects.select_related("target_population").get(id=pp_id)
+
+                    if program.status == Program.FINISHED:
+                        start_data = program.start_date
+                        end_data = program.end_date
+
+                        processing_with_finished_program(payment_plan, start_data, end_data)
```

Comment:
`processing_with_finished_program` could be run just in loop per Program - create ProgramCycle/adjust dates to Program's dates and assign this Program's all TP and PPlans to this ProgramCycle, right?

Reply:
yes, correct